packages/mobile/src/merchantPayment/MerchantPaymentScreen.tsx

```diff
@@ -0,0 +1,286 @@
+import { navigateBack, navigateHome } from 'src/navigator/NavigationService'
+import { Screens } from 'src/navigator/Screens'
+import { StackParamList } from 'src/navigator/types'
+import { getPassword, retrieveOrGeneratePepper } from 'src/pincode/authentication'
+import { useTokenInfoBySymbol } from 'src/tokens/hooks'
+import { fetchTokenBalances } from 'src/tokens/reducer'
+import { Currency } from 'src/utils/currencies'
+import { divideByWei } from 'src/utils/formatting'
+import { getContractKitAsync } from 'src/web3/contracts'
+import { isDekRegisteredSelector } from 'src/web3/selectors'
+
+type Props = StackScreenProps<StackParamList, Screens.MerchantPayment>
+
+function MerchantPaymentScreen({ route }: Props) {
+  const { params: routeParams } = route
+  const { t } = useTranslation()
+  const e164PhoneNumber = useSelector(e164NumberSelector)
+  const dispatch = useDispatch() as (...args: unknown[]) => void
+  const [sdkCharge, setSdkCharge] = useState<Charge | null>(null)
+  const [business, setBusiness] = useState<BusinessData | null>(null)
+  const isLoading = !business
+  const [amount, setAmount] = useState(new BigNumber(0))
+  const [submitting, setSubmitting] = useState(false)
+  const localCurrencyCode = useSelector(getLocalCurrencyCode)
+  const feeEstimates = useSelector(feeEstimatesSelector)
+  const tokenInfo = useTokenInfoBySymbol(Currency.Dollar)
+  const tokenAddress = tokenInfo?.address!
+  const feeEstimate = feeEstimates[tokenAddress]?.[FeeType.SEND]
+  const storedDekFee = feeEstimates[tokenAddress]?.[FeeType.REGISTER_DEK]
+  const localToFeeExchangeRate = useCurrencyToLocalAmount(new BigNumber(1), Currency.Dollar)
+  const isDekRegistered = useSelector(isDekRegisteredSelector) ?? false
+  const navigation = useNavigation()
+
+  const initCharge = useCallback(
+    async (kit: ContractKit) => {
+      if (!kit) {
+        throw new Error('Missing kit')
+      }
+
+      const chainHandler = new ContractKitTransactionHandler(kit)
+      const charge = new Charge(routeParams.apiBase, routeParams.referenceId, chainHandler, false)
+      try {
+        const info = await charge.getInfo()
+        setSdkCharge(charge)
+        setAmount(divideByWei(new BigNumber(info.action.amount)).decimalPlaces(2))
+        setBusiness(info.receiver.businessData)
+      } catch (e: unknown) {
+        dispatch(showError(t('merchantPaymentSetup')))
+        navigateBack()
+      }
+    },
+    [dispatch]
+  )
+
+  const submitPayment = useCallback(async () => {
+    if (sdkCharge && !submitting) {
+      setSubmitting(true)
+      try {
+        await sdkCharge.submit({ phoneNumber: e164PhoneNumber })
+        setSubmitting(false)
+        dispatch(showMessage(t('merchantPaymentSubmitSuccess')))
+        navigateHome()
+      } catch (e) {
+        setSubmitting(false)
+        dispatch(showError(t('merchantPaymentSubmitFail')))
+      }
+    }
+  }, [sdkCharge, dispatch, submitting])
+
+  useEffect(() => {
```

Comment:
For async operations we use `useAsync` instead of `useEffect`. It behaves very similarly with the difference that it allows the function to be `async` and returns a `loading`, `error` and `result` with the result of the async function.